### PR TITLE
VULCAN-26 VULCAN-32 VULCAN-41 Improve view managements and looks

### PR DESCRIPTION
**Pull request by ahdezma**

### CODE REVIEW

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt [hunk 1]
```diff
@@ -3,83 +3,114 @@ package com.liferay.vulcan.consumer.screens
...
 
-            (layout as? BaseView)?.screenlet = this
+    fun load(thingId: String, scenario: Scenario? = null, onComplete: ((ThingScreenlet) -> Unit)? = null) {
+        fetch(HttpUrl.parse(thingId)!!) {
+            if (scenario != null) {
+                this.scenario = scenario
+            }
 
-            it.failure { viewModel?.showError(it.message) }
+            thing = it.component1()
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt]: component1 is a very strange name

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt [hunk 2]
```diff
@@ -3,83 +3,114 @@ package com.liferay.vulcan.consumer.screens
...
 
             onComplete?.invoke(this)
         }
     }
 
+    private fun getLayoutIdFor(thing: Thing?): Int? {
+        if (layoutId != 0) return layoutId
+
+        return thing?.let {
+            onEventFor(GetLayoutEvent(thing = it, scenario = scenario))
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt]: GetLayout for a class name is also strange, it's like a method with the get notation

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt [hunk 3]
```diff
@@ -3,83 +3,114 @@ package com.liferay.vulcan.consumer.screens
...
-        reference = typedArray?.getResourceId(R.styleable.ThingScreenlet_layoutId, 0) ?: 0
-    }
+        layoutId = typedArray?.getResourceId(R.styleable.ThingScreenlet_layoutId, 0) ?: 0
+
+        val scenarioId = typedArray?.getString(R.styleable.ThingScreenlet_scenario) ?: ""
 
-    fun <T> onEventFor(action: Action<T>) = when (action) {
-        is ClickAction -> screenletEvents?.onClickEvent(layout as BaseView, action.view, action.thing)
+        scenario = Scenario.stringToScenario?.invoke(scenarioId) ?: when (scenarioId.toLowerCase()) {
+            "detail", "" -> Detail
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt]: this looks like a good situation for an enum

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/views/BaseView.kt [hunk 4]
```diff
@@ -1,27 +1,23 @@
...
 import com.liferay.vulcan.consumer.screens.ThingScreenlet
-import com.liferay.vulcan.consumer.screens.ViewModel
 import org.jetbrains.anko.AnkoLogger
 import org.jetbrains.anko.error
 
 /**
  * @author Alejandro Hernández
  */
-abstract class BaseView(context: Context, attrs: AttributeSet) : FrameLayout(context, attrs), ViewModel, AnkoLogger {
+interface BaseView : AnkoLogger {
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/views/BaseView.kt]: mmm :P coupling with anko

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/views/custom/PersonPortraitView.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    override var screenlet: ThingScreenlet? = null
+
+    val imageView by bindNonNull<ImageView>(R.id.image_view)
+
+    override var thing: Thing? by converter<Person> {
+        it.email
+            ?.trim()
+            ?.toLowerCase()
+            ?.let(String::md5)
```
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/views/custom/PersonPortraitView.kt]: ok, you are using it for this. I wouldn't define it in string

File: samples/blog/src/main/kotlin/com/liferay/vulcan/blog/postings/view/BlogPostingDetailCustom.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    val createDate by bindNonNull<TextView>(R.id.create_date)
+
+    override var thing: Thing? by converter<BlogPosting> {
+        headline.text = it.headline
+
+        alternativeHeadline.text = it.alternativeHeadline
+
+        Html.fromHtml(it.articleBody, Html.FROM_HTML_MODE_COMPACT)
+            .toString()
+            .replace("\n", "\n\n")
```
Comment (nhpatt) [inline samples/blog/src/main/kotlin/com/liferay/vulcan/blog/postings/view/BlogPostingDetailCustom.kt]: mmmm

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/views/custom/PersonPortraitView.kt (hunk 5 above)
Comment (ahdezma) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/views/custom/PersonPortraitView.kt]: Moved here: https://github.com/liferay-mobile/vulcan-consumer-android/pull/8/commits/49387093051602aebcbf60c97216809278af26b1

File: library/src/main/kotlin/com/liferay/vulcan/consumer/extensions/String+.kt [hunk 7]
```diff
@@ -1,9 +1,22 @@
...
 import java.text.SimpleDateFormat
 import java.util.Date
 import java.util.Locale
 
 fun String.asDate(format: SimpleDateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US)): Date? =
-    try { format.parse(this) } catch (parseException: ParseException) { null }
\ No newline at end of file
+    try { format.parse(this) } catch (parseException: ParseException) { null }
+
+fun String.md5(): String {
```
Comment (ahdezma) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/extensions/String+.kt]: Moved here: https://github.com/liferay-mobile/vulcan-consumer-android/pull/8/commits/49387093051602aebcbf60c97216809278af26b1

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt [hunk 8]
```diff
@@ -3,83 +3,114 @@ package com.liferay.vulcan.consumer.screens
...
+
+    var thing: Thing? by observe {
+        val layoutId = getLayoutIdFor(thing = it) ?: R.layout.thing_default
 
-    var thing: Thing? = null
-        set(value) {
-            field = value
-            viewModel?.thing = value
+        layout?.also {
+            this.removeView(it)
```
Comment (ahdezma) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt]: Kotlin alternative to:

```kotlin
if (layout != null) {
    this.removeView(layout)
}
```

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt (hunk 1 above)
Comment (ahdezma) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt]: They are standard for destructuring in Kotlin: https://kotlinlang.org/docs/reference/multi-declarations.html

File: library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt (hunk 8 above)
Comment (nhpatt) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/screens/ThingScreenlet.kt]: Yeah, I mean why removeView... are we removing in screens?

File: library/src/main/kotlin/com/liferay/vulcan/consumer/extensions/Byte+.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.liferay.vulcan.consumer.extensions
```
Comment (ahdezma) [inline library/src/main/kotlin/com/liferay/vulcan/consumer/extensions/Byte+.kt]: Moved here: 4938709

**Comment by nhpatt** [comment]
👍

**Pull request merged**